app/lib/storage/vault.ts

```diff
@@ -0,0 +1,121 @@
+import * as hydra from "hydration"
+import * as level from "level"
+import {LevelUp} from "levelup"
+import {homedir} from "os"
+import * as path from "path"
+import {Utils, Encryption} from "."
+
+/**
+ * Vaults are transparently encrypted LevelDB databases.
+ */
+export default class Vault {
+  /** Vault name */
+  protected name: string
+  /** Internal level (levelup+leveldown) reference */
+  protected db: LevelUp
+  /** Crypto options structure */
+  private cryptOptions: Encryption.Options
+  /** Vault can be used only after this promise is resolved */
+  public ready: Promise<boolean>
+
+  /**
+   * Vault constructor.
+   *
+   * WARNING: Vaults can't be used immediately after they are constructed
+   * because the constructor itself depends on asynchronous code. Please await
+   * for the `this.ready` promise before using any vault method.
+   *
+   * @param {string} name
+   * @param {string} password
+   */
+  constructor(name: string, password: string) {
+    this.name = name
+    const dbPath = path.normalize(`${homedir()}/.sheikah/vault/${name}`)
+
+    // This is the asynchronous part of the constructor.
+    this.ready = new Promise(async resolve => {
+      // Ensure the LevelDB directory exists
+      await Utils.ensurePath(dbPath)
+
+      // Create and set a new encryption options object with everything set to
+      // default but the password attribute, which is user defined.
+      this.cryptOptions = new Encryption.Options(
+        { password } as Encryption.IProtoOptions
+      )
+
+      // Create and set the underlying Level (levelup + leveldown) database.
+      this.db = level(dbPath, {
+        keyEncoding: "binary",
+        valueEncoding: "binary"
+      })
+
+      // This causes the `this.ready` promise to resolve.
+      resolve(true)
+    })
+  }
+
+  /**
+   * Create or overwrite a value entry for a certain key.
+   * @param {string} key
+   * @param value
+   * @returns {Promise<boolean>}
+   */
+  public async put(key: string, value: any): Promise<boolean> {
+    // Keys are hashed for privacy and to allow arbitrary key size
+    const hashedKey = Encryption.keyHash(key)
+    // Dehydrate/deflate and serialize the object to convert it into a typed
+    // JSON that can be rehydrated/inflated afterwards without losing types.
+    const serializedValue = Buffer.from(JSON.stringify(hydra.dehydrate(value)))
+    // Encrypt the serialized object.
+    const encryptedValue =
+      Encryption.encrypt(serializedValue, this.cryptOptions)
+    // Put the value into the database and wait for write to complete.
+    await new Promise(resolve => {
+      this.db.put(hashedKey, encryptedValue, resolve)
+    })
+
+    return true
+  }
```

Comment:
Couldn't/shouldn't this case be handled as in the `get` function? `put` method can also give errors:
```
      this.db.put(hashedKey, encryptedValue, (error, value) => {
        if (error) {
          reject(error)
        } else {
          resolve(true)
        }
      })
```